subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/NormalizedPathFingerprintCompareStrategy.java

```diff
@@ -56,61 +50,144 @@ private NormalizedPathFingerprintCompareStrategy() {
      *         For those only in the previous fingerprint collection it checks if some entry with the same normalized path is in the current collection.
      *         If it is, file is reported as modified, if not as removed.
      *     </li>
-     *     <li>Finally, if {@code includeAdded} is {@code true}, the remaining fingerprints which are only in the current collection are reported as added.</li>
+     *     <li>Finally, if {@code shouldIncludeAdded} is {@code true}, the remaining fingerprints which are only in the current collection are reported as added.</li>
      * </ul>
      */
     @Override
-    protected boolean doVisitChangesSince(ChangeVisitor visitor, Map<String, FileSystemLocationFingerprint> currentFingerprints, Map<String, FileSystemLocationFingerprint> previousFingerprints, String propertyTitle, boolean includeAdded) {
-        ListMultimap<FileSystemLocationFingerprint, FilePathWithType> unaccountedForPreviousFiles = MultimapBuilder.hashKeys(previousFingerprints.size()).linkedListValues().build();
-        ListMultimap<String, FilePathWithType> addedFilesByNormalizedPath = MultimapBuilder.linkedHashKeys().linkedListValues().build();
-        for (Map.Entry<String, FileSystemLocationFingerprint> entry : previousFingerprints.entrySet()) {
+    protected boolean doVisitChangesSince(
+        ChangeVisitor visitor,
+        Map<String, FileSystemLocationFingerprint> currentFingerprints,
+        Map<String, FileSystemLocationFingerprint> previousFingerprints,
+        String propertyTitle,
+        boolean shouldIncludeAdded
+    ) {
+        ListMultimap<FileSystemLocationFingerprint, FilePathWithType> missingPreviousFiles = getMissingPreviousFingerprints(previousFingerprints);
+        ListMultimap<String, FilePathWithType> addedFilesByNormalizedPath = getAddedFilesByNormalizedPath(currentFingerprints, missingPreviousFiles);
+
+        for (Entry<FileSystemLocationFingerprint, FilePathWithType> entry : missingPreviousFiles.entries()) {
+            if (wasModifiedOrRemovedAndMessageCountSaturated(visitor, propertyTitle, addedFilesByNormalizedPath, entry)) {
+                return false;
+            }
+        }
+
+        if (shouldIncludeAdded) {
+            for (Entry<String, FilePathWithType> entry : addedFilesByNormalizedPath.entries()) {
+                if (wasAddedAndMessageCountSaturated(visitor, propertyTitle, entry)) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    private ListMultimap<FileSystemLocationFingerprint, FilePathWithType> getMissingPreviousFingerprints(
+        Map<String, FileSystemLocationFingerprint> previousFingerprints
+    ) {
+        ListMultimap<FileSystemLocationFingerprint, FilePathWithType> results = MultimapBuilder
+            .treeKeys()
```

Comment:
Using a tree set as entry set may cause quite a performance regression, since insertion sort is quadratic. Either build an immutable multimap or sort the missing fingerprints as before.

Reply:
It uses a standard TreeMap inside: https://google.github.io/guava/releases/21.0/api/docs/src-html/com/google/common/collect/MultimapBuilder.html#line.168
It has the same complexity for inserting n elements: nlog(n):
> https://en.wikipedia.org/wiki/Red%E2%80%93black_tree
The insertion and deletion operations, along with the tree rearrangement and recoloring, are also performed in O(log n) time.

Reply:
If you want to simplify the code, I think the best option is to keep the `hashKeys()`, and go for Java 8:

```java
missingPreviousFiles.entries().stream()
    .sorted(Map.Entry.comparingByKey())
    .forEach(/* ... */)
```

Reply:
Oh, awesome, didn't know I can use Java 8 here :)

Reply:
Jaba 8 _API_ is available most everywhere except for a small number of very specific places (worker daemons, bootstrapping logic etc.). Java 8 _syntax_ is not available in some of the subprojects (including `:core`).

Here you can use Java 8 API with Java 8 syntax. In most of `:core` and some other projects you can use Java 8 API, but you have to use Java 6 syntax. Doing the latter makes the code look ugly, but might pay off still in some cases.

Once we separate the Java 6-compatible code, we can use Java 8 syntax and API everywhere else.